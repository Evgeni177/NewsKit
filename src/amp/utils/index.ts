export * from './render';
export * from './style-tag';
