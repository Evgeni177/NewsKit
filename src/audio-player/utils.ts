export const formatTrackData = (
  trackColor: string,
  indicatorColor: string,
  bufferColor: string,
  timeArr: number[],
  buffered: TimeRanges | undefined,
) => {
  if (!buffered) {
    return {colors: [], values: []};
  }

  const time = timeArr[0];
  const bufferPositions = [];

  const {length} = buffered;
  for (let i = 0; i < length; i += 1) {
    const startPosition = buffered.start(i);
    if (startPosition > time) {
      bufferPositions.push(startPosition);
    }

    const endPosition = buffered.end(i);
    if (endPosition > time) {
      bufferPositions.push(endPosition);
    }
  }

  const colors = [
    indicatorColor,
    ...bufferPositions.map((x, i) => (i % 2 ? trackColor : bufferColor)),
    trackColor,
  ];
  const values = [time, ...bufferPositions];

  return {colors, values};
};
