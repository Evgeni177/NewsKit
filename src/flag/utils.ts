export enum FlagSize {
  Small = 'small',
  Large = 'large',
}
