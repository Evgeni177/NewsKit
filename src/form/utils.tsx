import React from 'react';
import {FormProps, FormInputState} from './types';
import {IconFilledCheckCircle, IconFilledError} from '../icons';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const composeEventHandlers = (listeners: any[] = []) => (event: any) => {
  listeners.forEach(listener => {
    if (typeof listener === 'function') {
      listener(event);
    }
  });
};

export const getStatusIcon = ({
  state,
  validIcon,
  invalidIcon,
  iconSize,
}: {
  state?: FormInputState;
  validIcon?: React.ReactNode;
  invalidIcon?: React.ReactNode;
  iconSize: string;
}): React.ReactNode => {
  if (state === 'invalid') {
    return (
      invalidIcon || (
        <IconFilledError
          data-testid="error-icon"
          overrides={{
            size: iconSize,
            stylePreset: 'iconNegative',
          }}
        />
      )
    );
  }
  if (state === 'valid') {
    return (
      validIcon || (
        <IconFilledCheckCircle
          data-testid="tick-icon"
          overrides={{
            size: iconSize,
            stylePreset: 'iconPositive',
          }}
        />
      )
    );
  }

  return null;
};

export const excludeReactHookFormProps = (props: FormProps) => {
  const {
    resolver,
    defaultValues,
    reValidationMode,
    validationMode,
    ...rest
  } = props;

  return rest;
};
