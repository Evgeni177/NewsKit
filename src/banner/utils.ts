import {BreakpointKeys, Theme} from '../theme';
import {MQ} from '../utils';
import {sortBreakpointKeys} from '../utils/hooks/use-media-query/utils';
import {BannerInternalProps} from './types';

const isVertical = (layout?: BannerInternalProps['layout']) =>
  layout === 'vertical';

export const getVisibleBreakpointsForLayout = (
  layout: Partial<
    {
      [key in BreakpointKeys]: 'horizontal' | 'vertical';
    }
  >,
  theme: Theme,
) => {
  const horizontalBreakpoints: MQ<boolean> = {};
  const verticalBreakpoints: MQ<boolean> = {};

  const {breakpoints} = theme;
  const breakpointsKeys = Object.keys(breakpoints) as BreakpointKeys[];

  // Get the lowest breakpoint key as base
  const baseBreakpoint: BreakpointKeys = sortBreakpointKeys(
    Object.keys(layout) as BreakpointKeys[],
    breakpointsKeys,
  )[0];
  let baseValue = layout[baseBreakpoint];

  // Set values to all breakpoints (mainly for the missing ones)
  const valuesForAllBreakpoints = breakpointsKeys.reduce(
    (acc, breakpointKey) => {
      if (acc[breakpointKey] === undefined) {
        return {...acc, [breakpointKey]: baseValue};
      }

      baseValue = acc[breakpointKey];
      return acc;
    },
    layout,
  );

  const valuesForAllBreakpointsAsArr = Object.entries(valuesForAllBreakpoints);

  // Creating breakpoint objects like:
  // verticalBreakpoints = {
  //   xs: true,
  //   sm: true,
  //   md: false,
  //   lg: false,
  //   xl: false,
  // }
  // That can be passed to the visible component based on the layout
  valuesForAllBreakpointsAsArr.forEach(([key, value]) => {
    verticalBreakpoints[key as BreakpointKeys] = isVertical(value);
    horizontalBreakpoints[key as BreakpointKeys] = !isVertical(value);
  });

  return {
    verticalBreakpoints,
    horizontalBreakpoints,
  };
};
