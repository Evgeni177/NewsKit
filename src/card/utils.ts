import {getStylePreset} from '../utils/style';

export const isHorizontal = (layout?: string) =>
  layout && layout.includes('horizontal');

export const isReverse = (layout?: string) =>
  layout && layout.includes('reverse');

export const filterInteractiveStates = (path: string, hasHref?: boolean) =>
  getStylePreset(
    `card${path ? `.${path}` : ''}`,
    path,
    hasHref ? {} : {omitStates: ['hover', 'active']},
  );
