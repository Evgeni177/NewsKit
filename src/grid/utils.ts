import {BreakpointKeys} from '../theme';
import {ThemeProp} from '../utils/style-types';
import {GridProps} from './types';

export enum OverrideProp {
  Margin = 'Margin',
  ColumnGutter = 'ColumnGutter',
  RowGutter = 'RowGutter',
}

const propThemeMap = {
  [OverrideProp.Margin]: 'containerMargin',
  [OverrideProp.ColumnGutter]: 'columnGutters',
  [OverrideProp.RowGutter]: 'rowGutters',
};

export const getOverridableProp = (
  prop: OverrideProp,
  breakpoint: BreakpointKeys,
  {theme, ...props}: GridProps & ThemeProp,
) => {
  const breakpoints = Object.keys(theme.breakpoints);
  const gutterOverrideToken = breakpoints
    .slice(0, breakpoints.indexOf(breakpoint) + 1)
    .reverse()
    .map(bkp => props[`${bkp}${prop}` as keyof GridProps])
    .filter(Boolean)[0];

  const themeKey = propThemeMap[prop];

  // Gutter from theme is a px number, but override is a string token, so convert it to px number.
  return gutterOverrideToken
    ? parseInt(theme.sizing[gutterOverrideToken], 10)
    : (theme.componentDefaults.grid[themeKey] as Record<
        BreakpointKeys,
        number
      >)[breakpoint];
};
