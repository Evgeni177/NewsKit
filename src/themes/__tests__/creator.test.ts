import {createTheme, newskitLightTheme} from '..';
import {styled} from '../../utils/style';
import {getStylePreset} from '../../utils/style-preset';
import {renderWithTheme} from '../../test/test-utils';

describe('themeing functions', () => {
  describe('createTheme', () => {
    test('override default presets with existing preset ', () => {
      const theme = createTheme('newskit-light-demo', {
        baseTheme: newskitLightTheme,
        themeOverrider: () => ({
          defaultPresets: {
            headline: {
              heading: {
                stylePreset: 'buttonSolidPrimary',
              },
            },
          },
        }),
      });

      expect(theme).not.toEqual(newskitLightTheme);
      expect(theme.defaultPresets.headline.heading.stylePreset).toEqual(
        'buttonSolidPrimary',
      );
    });

    test('override default presets with newly created preset ', () => {
      const theme = createTheme('newskit-light-demo', {
        baseTheme: newskitLightTheme,
        themeOverrider: primitives => ({
          stylePresets: {
            buttonTestStyle: {
              base: {
                color: primitives.colors.red100,
                backgroundColor: primitives.colors.red100,
              },
            },
          },
          defaultPresets: {
            headline: {
              heading: {
                stylePreset: 'buttonTestStyle',
              },
            },
          },
        }),
      });

      const TestSurface = styled.div`
        display: inline-block;
        ${getStylePreset('headline.heading', 'heading')}
      `;

      const {asFragment} = renderWithTheme(TestSurface, undefined, theme);
      expect(asFragment()).toMatchSnapshot();
    });

    test('create default Theme', () => {
      const theme = createTheme('newskit-light');
      const copy = Object.assign({}, newskitLightTheme);
      const expected = Object.assign(copy, {name: 'newskit-light'});
      expect(theme).toEqual(expected);
    });

    test('override by passing newskitLightTheme', () => {
      const theme = createTheme('newskit-light', {
        baseTheme: newskitLightTheme,
      });
      expect(theme).toEqual(newskitLightTheme);
    });

    test('override using theme overriding function', () => {
      const theme = createTheme('newskit-light', {
        baseTheme: newskitLightTheme,
        themeOverrider: () => ({
          colors: {
            black: 'nooop',
          },
        }),
      });
      expect(theme).not.toEqual(newskitLightTheme);
      expect((theme.colors as any).black).toEqual('nooop');
    });

    test('override of color is propagated in styledPreset', () => {
      const testColor = '#eeff99';
      const theme = createTheme('newskit-light', {
        colorOverrides: {
          interactive030: testColor,
        },
        baseTheme: newskitLightTheme,
      });
      expect(theme.colors.interactive030).toEqual(testColor);
      expect(
        theme.stylePresets.iconButtonSolidPrimary &&
          theme.stylePresets.iconButtonSolidPrimary.base &&
          theme.stylePresets.iconButtonSolidPrimary.base.backgroundColor,
      ).toEqual(testColor);
    });

    test('override of stylePreset is present in theme', () => {
      const customStylePreset = {
        base: {
          backgroundColor: '#101010',
          borderRadius: '20rem',
        },
        hover: {
          backgroundColor: '#909090',
        },
      };
      const theme = createTheme('newskit-light', {
        themeOverrider: () => ({
          stylePresets: {
            audioPlayerSeekBarIndicator: customStylePreset,
          },
        }),
      });
      expect(theme.stylePresets.audioPlayerSeekBarIndicator).toEqual(
        customStylePreset,
      );
    });
  });
});
