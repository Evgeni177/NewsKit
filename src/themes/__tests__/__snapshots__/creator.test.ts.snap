// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`themeing functions createTheme override default presets with newly created preset  1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-block;
  color: #4d0000;
  background-color: #4d0000;
}

<div
    class="emotion-0"
  />
</DocumentFragment>
`;
